File: python-scraper/tiktok.py
import os
import requests
import json
from concurrent.futures import ThreadPoolExecutor
from dotenv import load_dotenv

load_dotenv()

def get_posts(username,  maxResults, with_videos=False):
    if not os.getenv('APIFY_API_KEY'):
        raise Exception("APIFY_API_KEY is not set in the environment variables")

    data = {
        "excludePinnedPosts": False,
        "profiles": [username],
        "resultsPerPage": maxResults,
        "shouldDownloadCovers": False,
        "shouldDownloadSlideshowImages": False,
        "shouldDownloadSubtitles": False,
        "shouldDownloadVideos": with_videos,
        "searchSection": "",
        "maxProfilesPerQuery": 1,
    }

    headers = {
        "Authorization": f"Bearer {os.getenv('APIFY_API_KEY')}"
    }

    try:
        response = requests.post(
            "https://api.apify.com/v2/acts/clockworks~free-tiktok-scraper/run-sync-get-dataset-items",
            json=data,
            headers=headers
        )

        if response.status_code != 201:
            raise Exception("Failed to fetch posts")

        print("RETURNED DATA", response.json(), response.status_code)

        returned_posts = response.json()

        cleaned_posts = [
            {
                "id": post["id"],
                "timestamp": post["createTime"],
                "postUrl": post["webVideoUrl"],
                "videoUrl": post.get("videoMeta", {}).get("downloadAddr", ""),
                "audioUrl": post["musicMeta"]["playUrl"],
                "caption": post["text"],
                "hashtags": post["hashtags"],
                "isPinned": post["isPinned"],
                "isSponsored": post["isSponsored"],
            }
            for post in returned_posts
        ]

        print(cleaned_posts)

        return cleaned_posts
    except Exception as error:
        print("Error fetching posts:", error)
        raise

def download_tt_posts(username, posts, with_videos):
    folder_name = f"{username}_tt_media"
    audio_folder_name = os.path.join(folder_name, "audios")
    video_folder_name = os.path.join(folder_name, "videos")

    os.makedirs(audio_folder_name, exist_ok=True)
    os.makedirs(video_folder_name, exist_ok=True)

    def download_post(post):
        try:
            if post["audioUrl"]:
                response = requests.get(post["audioUrl"])
                file_name = os.path.join(audio_folder_name, f"tt_{username}_{post['id']}_audio.mp3")
                print(f"Saving {file_name}")
                with open(file_name, 'wb') as f:
                    f.write(response.content)

            if with_videos and post["videoUrl"]:
                response = requests.get(post["videoUrl"])
                file_name = os.path.join(video_folder_name, f"tt_{username}_{post['id']}_video.mp4")
                print(f"Saving {file_name}")
                with open(file_name, 'wb') as f:
                    f.write(response.content)
        except Exception as error:
            print(f"Error downloading video for post {post['id']}:", error)

    with ThreadPoolExecutor(max_workers=2) as executor:
        executor.map(download_post, posts)

if __name__ == "__main__":
    username = "willcaplan_changemakers"
    print("fetching profile data for...", username)
    with_videos = False
    maxResults = 2
    try:
        posts = get_posts(username, maxResults, with_videos)
        with open(f"tt_{username}_posts.json", 'w') as f:
            json.dump(posts, f, indent=2)

        download_tt_posts(username, posts, with_videos)
    except Exception as error:
        print("An error occurred:", error)

File: typescript-scraper/tiktok.ts
import axios from "axios";
import fs from "fs";
import Bluebird from "bluebird";
import dotenv from "dotenv";
dotenv.config();

const getPosts = async ({
  username,
  maxResults,
  withVideos = false,
}: {
  username: string;
  maxResults: number;
  withVideos: boolean;
}) => {
  if (!process.env.APIFY_API_KEY) {
    throw new Error("APIFY_API_KEY is not set in the environment variables");
  }

  const data = {
    excludePinnedPosts: false,
    profiles: [username],
    resultsPerPage: maxResults,
    shouldDownloadCovers: false,
    shouldDownloadSlideshowImages: false,
    shouldDownloadSubtitles: false,
    shouldDownloadVideos: withVideos,
    searchSection: "",
    maxProfilesPerQuery: 1,
  };

  const headers = {
    headers: {
      Authorization: `Bearer ${process.env.APIFY_API_KEY}`,
    },
  };

  try {
    const response = await axios.post(
      `https://api.apify.com/v2/acts/clockworks~free-tiktok-scraper/run-sync-get-dataset-items`,
      data,
      headers
    );

    if (response.status !== 201) {
      throw new Error("Failed to fetch posts");
    }

    console.log("RETURNED DATA", response.data, response.status);

    const returnedPosts = response.data;

    const cleanedPosts: CleanedPost[] = returnedPosts.map(
      (post: TikTokPost) => ({
        id: post.id,
        timestamp: post.createTime,
        postUrl: post.webVideoUrl,
        videoUrl: post?.videoMeta?.downloadAddr ?? "",
        audioUrl: post.musicMeta.playUrl,
        caption: post.text,
        hashtags: post.hashtags,
        isPinned: post.isPinned,
        isSponsored: post.isSponsored,
      })
    );

    console.log(cleanedPosts);

    return cleanedPosts;
  } catch (error) {
    console.error("Error fetching posts:", error);
    throw error;
  }
};

const downloadTTPosts = async ({
  username,
  posts,
  withVideos,
}: {
  username: string;
  posts: CleanedPost[];
  withVideos: boolean;
}) => {
  const folderName = `${username}_tt_media`;

  const audioFolderName = `${folderName}/audios`;

  if (!fs.existsSync(audioFolderName)) {
    await fs.promises.mkdir(audioFolderName, { recursive: true });
  }

  const videoFolderName = `${folderName}/videos`;

  if (!fs.existsSync(videoFolderName)) {
    await fs.promises.mkdir(videoFolderName, { recursive: true });
  }

  await Bluebird.map(
    posts,
    async (post) => {
      try {
        if (post.audioUrl) {
          const response = await axios.get(post.audioUrl, {
            responseType: "arraybuffer",
          });
          const buffer = Buffer.from(response.data, "binary");
          const fileName = `${audioFolderName}/tt_${username}_${post.id}_audio.mp3`;
          console.log(`Saving ${fileName}`);
          await fs.promises.writeFile(fileName, buffer);
        }

        if (withVideos && post.videoUrl) {
          const response = await axios.get(post.videoUrl, {
            responseType: "arraybuffer",
          });
          const buffer = Buffer.from(response.data, "binary");
          const fileName = `${videoFolderName}/tt_${username}_${post.id}_video.mp4`;
          console.log(`Saving ${fileName}`);
          await fs.promises.writeFile(fileName, buffer);
        }
      } catch (error) {
        console.error(`Error downloading video for post ${post.id}:`, error);
      }
    },
    { concurrency: 2 }
  );
};

(async () => {
  const username = "willcaplan_changemakers";
  console.log("fetching profile data for...", username);
  const withVideos = false;
  const maxResults = 200;
  try {
    const posts = await getPosts({ username, maxResults, withVideos });
    fs.writeFileSync(
      `tt_${username}_posts.json`,
      JSON.stringify(posts, null, 2)
    );

    await downloadTTPosts({ username, posts, withVideos });
  } catch (error) {
    console.error("An error occurred:", error);
  }
})();

interface TikTokPost {
  id: string;
  text: string;
  createTime: number;
  createTimeISO: string;
  authorMeta: {
    id: string;
    name: string;
    profileUrl: string;
    nickName: string;
    verified: boolean;
    signature: string;
    bioLink: string | null;
    avatar: string;
    commerceUserInfo: object;
    privateAccount: boolean;
    region: string;
    roomId: string;
    ttSeller: boolean;
    following: number;
    friends: number;
    fans: number;
    heart: number;
    video: number;
    digg: number;
  };
  musicMeta: {
    musicName: string;
    musicAuthor: string;
    musicOriginal: boolean;
    playUrl: string;
    coverMediumUrl: string;
    musicId: string;
  };
  webVideoUrl: string;
  mediaUrls: string[];
  videoMeta?: {
    height: number;
    width: number;
    duration: number;
    coverUrl: string;
    originalCoverUrl: string;
    definition: string;
    format: string;
    subtitleLinks: any[];
    downloadAddr: string;
  };
  diggCount: number;
  shareCount: number;
  playCount: number;
  collectCount: number;
  commentCount: number;
  mentions: any[];
  hashtags: object[];
  effectStickers: any[];
  isSlideshow: boolean;
  isPinned: boolean;
  isSponsored: boolean;
  input: string;
}

interface CleanedPost {
  id: string;
  timestamp: number;
  postUrl: string;
  videoUrl?: string;
  audioUrl: string;
  caption: string;
  hashtags: object[];
  isPinned: boolean;
  isSponsored: boolean;
}
